README.md
# A tool for building self contained Lua executables

This tool is an alternative for ```srlua``` which can be found [here](http://tecgraf.puc-rio.br/~lhf/ftp/lua/).

The main points where ```dogfood``` differs from ```srlua``` are:

* One self contained executable.
* Can embed multiple Lua modules in one executable.
* No support for Lua 5.1 and older.

Other features:

* Does not depend on a C compiler to build self contained Lua executables.
* ```dogfood``` self can be build for Unix like operating systems _and_ Windows.

```dogfood``` consists of two parts; a Lua interpreter and a Lua module that creates the self executable Lua programs.
The interpreter runs the dogfood's Lua module that is appended at the end of the executable.
The module reuses the interpreter when it builds a self contained Lua executable by copying its interpreter and then append the user's Lua modules.

```dogfood``` owes its name to the reuse of its own interpreter.
Using your own software products is also knowns as [dogfooding](https://en.wikipedia.org/wiki/Eating_your_own_dog_food).

## Requirements

* Lua version 5.2, 5.3 or 5.4

## Usage of dogfood

``` sh
dogfood [OPTIONS] OUT MODULE...
```

```OUT``` is the destination of the resulting program  
The first ```MODULE``` from the list is the entry point of the program while the optional extra modules are accessable by ```require```.
The modules are provided _without_ the '.lua' extention.  
C modules cannot be embedded in the resulting program but the program can load C modules when they are placed in one of the ```package.cpath``` search paths.

The following ```OPTIONS``` are available.

|Option | Description|
|-------|------------|
|-c, --compile | Compile the modules and embed them as bytecode.|
|-s, --strip-debug-information | Strips the debug information from the bytecode.|
|-h, --help | Shows the help.|
|-m | Adds the given path to ```package.loaded``` to search for modules that are provided as parameter.|
|-v, --lua-version | Shows the Lua language version of the interpreter used by the resulting program..|

## Dogfood binaries for Windows

Building dogfood binaries for Windows requires more effort than for Unix like operating systems.  
The following executables are available for download to lower the bar for Windows users that want to try ```dogfood``` or don't have the knowledge to build it.

| Lua version | dogfood v1.1.0 x86-64 | dogfood v1.1.0 x86-32 |
|-------------|-----|-----|
| 5.2.4 | [download](https://raw.githubusercontent.com/PG1003/dogfood/master/exe/dogfood52_win_x86-64.zip) | [download](https://raw.githubusercontent.com/PG1003/dogfood/master/exe/dogfood52_win_x86-32.zip) |
| 5.3.6 | [download](https://raw.githubusercontent.com/PG1003/dogfood/master/exe/dogfood53_win_x86-64.zip) | [download](https://raw.githubusercontent.com/PG1003/dogfood/master/exe/dogfood53_win_x86-32.zip) |
| 5.4.6 | [download](https://raw.githubusercontent.com/PG1003/dogfood/master/exe/dogfood54_win_x86-64.zip) | [download](https://raw.githubusercontent.com/PG1003/dogfood/master/exe/dogfood54_win_x86-32.zip) |
                              
## Bootstrapping

Right after compilation, a dogfood binary must be bootstrapped before it is able to build self contained Lua executables.
The makefile provided with the sources already takes care of this.
In case when you have your own build environment, a bootstrap shell script is provided for Unix like environments and for Windows.

The usage of the bootstrap script is as follows;

``` sh
bootstrap.sh EXECUTABLE food.lua OUT
```

For Windows ```bootstrap.bat``` instead of ```bootstrap.sh``` is used but the commandline prameters are same.

```EXECUTABLE``` is the path to the executable that needs to be bootstrapped.  
The file ```food.lua``` contains dogfood's logic.
You can find this file in the [source](/src) directory.  
```OUT``` is the destination of the resulting executable.
```EXECUTABLE``` and ```OUT``` can point to the same file.


CHANGELOG.txt
# 1.1.0

Changes:
- Added for Windows the option to compile dogfood with support for an embed Lua DLL
  in the executable's resource.
  This is enabled by setting a EMBED_LUA_DLL preprocessor define.
  Details about embedding the DLL can be found in dog.c.
- Refactored to remove duplicate code and to reduce and simplify buffering while
  reading Lua data from executable.

# 1.0.2

Changes:
- Use lua_pcall when calling the main module and report errors to the user.
- Output all errors to stderr like the PUC Lua interpreter does.
- Output a warning for C modules that are found at one of package.cpath locations.

# 1.0.1

Changes:
- Added the '-v' and '--lua-version' options to show the Lua language version.

# v1.0.0

- Initial release


test/bar.lua
 
local bar = {}

function bar.bar()
    return "bar"
end

return bar


test/foo.lua
local bar    = require( "bar" )
local baz    = require( "bar.baz" )
local pg1003 = require( "pg1003" )

-- Test if we get the expected arguments
assert( string.match( arg[ 0 ], "/foobar$" ) )
assert( arg[ 1 ] == "-param" )

-- Call functions from other modules
assert( bar.bar() == "bar" )
assert( baz.baz() == "baz" )
assert( pg1003.pg1003() == "PG1003" )

-- Retun success
return 0


test/pg1003.c
#include "lua.h"
#include "lauxlib.h"

static int pg1003( lua_State *L )
{
    lua_checkstack( L, 1 );
    lua_pushliteral( L, "PG1003" );

    return 1;
}

static const struct luaL_Reg pg1003_functions[] =
{
    { "pg1003", pg1003 },
    { NULL, NULL }
};

LUALIB_API int luaopen_pg1003( lua_State *L )
{
    luaL_newlib( L,  pg1003_functions );

    return 1;
}


src/dog.c
// Copyright (c) 2019 PG1003
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to
// deal in the Software without restriction, including without limitation the
// rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
// sell copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.

#include <stdio.h>
#include <stdlib.h>
#include <stdarg.h>

#include <lua.h>
#include <lualib.h>
#include <lauxlib.h>

#if defined( _WIN32 ) && defined( EMBED_LUA_DLL )
    #define LOAD_EMBEDDED_LUA_DLL

    #include "resource.h"
    #include <windows.h>
    #include <tchar.h>
    #include <assert.h>
#endif

#define STR_HELPER(x) #x
#define STR(x) STR_HELPER(x)

#define MAX_NAME_LENGTH     127
#define MAX_NAME_LENGTH_STR STR( MAX_NAME_LENGTH )


#if LUA_VERSION_NUM < 502 || LUA_VERSION_NUM > 504
    #error "Unsupported Lua version"
#endif

#if defined( LOAD_EMBEDDED_LUA_DLL )

/* This function loads Lua DLL from its resource.
 * Embedding the Lua DLL simplifies the distribution of dogfood and its derived executables.
 *
 * This feature is specific for Windows and is enabled by a EMBED_LUA_DLL preprocessor define.
 *
 * The function requires that:
 * 1) The data is a valid Lua DLL that targets the same platform as dogfood's executable.
 * 2) The DLL is delay loaded.
 * 3) The ID for DLL in the resource is named IRD_LUADLL.
 * 4) The resource type of the Lua DLL is RCDATA.
 *
 * The resource file should contain a line that look like the following;
 *
 *    IDR_LUADLL    RCDATA    "PATH_TO_YOUR\Lua.dll"
 *
 * Its possible to link dogfood static with Lua but then support for C mudules will be effectively
 * disabled. C modules will likely require a Lua DLL that conflicts with the statically linked Lua
 * instance in dogfood. In this case a 'multiple Lua VMs detected' message will be displayed in
 * the console output when you run the resulting executable.
 */
static void load_lua_dll_from_resouce()
{
    /* https://stackoverflow.com/questions/17774103/using-an-embedded-dll-in-an-executable
     * https://learn.microsoft.com/en-us/cpp/windows/how-to-include-resources-at-compile-time */

#pragma warning( push, 3 )
    const HRSRC resource_handle = FindResource( NULL,
                                                MAKEINTRESOURCE( IDR_LUADLL ),
                                                RT_RCDATA );
    assert( resource_handle );
#pragma warning( pop )

    const HGLOBAL file_resouce_handle = LoadResource( NULL, resource_handle );
    assert( file_resouce_handle );

    const LPVOID file_at_resouce = LockResource( file_resouce_handle );
    assert( file_at_resouce );

    const DWORD  file_size = SizeofResource( NULL, resource_handle );

    const TCHAR  lua_dll_name[] = _T("Lua" LUA_VERSION_MAJOR LUA_VERSION_MINOR ".dll");
    const HANDLE file_handle    = CreateFile( lua_dll_name,
                                              GENERIC_READ | GENERIC_WRITE,
                                              0,
                                              NULL,
                                              CREATE_ALWAYS,
                                              FILE_ATTRIBUTE_NORMAL,
                                              NULL );
    assert( file_handle );

    const HANDLE file_mapping_handle = CreateFileMapping( file_handle,
                                                          NULL,
                                                          PAGE_READWRITE,
                                                          0,
                                                          file_size,
                                                          NULL );
    assert( file_mapping_handle );

    const LPVOID mapped_file_address = MapViewOfFile( file_mapping_handle,
                                                      FILE_MAP_WRITE,
                                                      0, 0, 0 );
    assert( mapped_file_address );

    CopyMemory( mapped_file_address, file_at_resouce, file_size );

    UnmapViewOfFile( mapped_file_address );
    CloseHandle( file_mapping_handle );
    CloseHandle( file_handle );
}

#endif

static void dogfood_error( lua_State * L, const char * const format, ... )
{
    va_list args;
    va_start( args, format );

    fputs( "Dogfood error: ", stderr );
    vfprintf( stderr, format, args );
    fputc( '\n', stderr );

    lua_close( L );
    exit( 1 );
}

static void dogfood_errno( lua_State * L )
{
    perror( "Dogfood error" );
    lua_close( L );
    exit( 1 );
}

typedef char module_name_buffer[ MAX_NAME_LENGTH + 1 ];

static void read_module_header( lua_State * L,
                                FILE * f,
                                module_name_buffer module_name,
                                unsigned long * module_size )
{
    const char header[]        = "\r\n-- %" MAX_NAME_LENGTH_STR "s %08lX%c%c";
    char       carriage_return = 0;
    char       new_line        = 0;

    const int result = fscanf( f, header,
                               module_name, module_size, &carriage_return, &new_line );

    if( result == EOF && ferror( f ) )
    {
        dogfood_errno( L );
    }

    if( result != 4 ||
        module_size == 0 ||
        carriage_return != '\r' ||
        new_line != '\n' )
    {
        dogfood_error( L, "No valid start of module '%s' found.", module_name );
    }
}

typedef struct
{
    FILE *       f;
    const char * module_name;
    size_t       remaining;
    size_t       available;
    char         buffer[ 4096 ];
} module_reader_ctx;

static const char * module_reader( lua_State *L, void * p, size_t * size )
{
    module_reader_ctx * const ctx = ( module_reader_ctx * )p;

    if( ctx->remaining == 0u )
    {
        return NULL;
    }

    const size_t max_buffer = sizeof( ctx->buffer );
    const size_t n_to_read  = ctx->remaining < max_buffer ? ctx->remaining : max_buffer;

    *size = fread( ctx->buffer, sizeof( char ), n_to_read, ctx->f );
    if( ferror( ctx->f ) )
    {
        dogfood_errno( L );
    }

    if( *size != n_to_read )
    {
        dogfood_error( L, "Cannot read the entire module '%s'.", ctx->module_name );
    }

    ctx->remaining -= *size;

    return ctx->buffer;
}

static void load_module( lua_State * L,
                         FILE * const f,
                         module_name_buffer module_name,
                         size_t module_size )
{
    module_reader_ctx ctx;
    ctx.f           = f;
    ctx.module_name = module_name;
    ctx.remaining   = module_size;
    ctx.available   = 0u;

    const int result = lua_load( L, module_reader, &ctx, module_name, NULL );
    switch( result )
    {
        case LUA_OK:
        case LUA_YIELD:
            break;

        case LUA_ERRSYNTAX:
        case LUA_ERRMEM:
#if LUA_VERSION_NUM < 504
        case LUA_ERRGCMM:
#endif
        case LUA_ERRERR:
        {
            const char * const extra = lua_tostring( L, -1 );
            dogfood_error( L,
                           extra ? "Error while loading module '%s';\n%s" :
                                   "An error has occurred while loading module '%s'.",
                           module_name, extra  );
        }
    }
}

int main( int argc, char *argv[] )
{
#if defined( LOAD_EMBEDDED_LUA_DLL )
    load_lua_dll_from_resouce();
#endif

    lua_State * const L = luaL_newstate();
    luaL_openlibs( L );

    /* Create and fill argument table */
    lua_createtable( L, argc, 0 );
    int i = 0;
    for( ; i < argc ; ++i )
    {
        lua_pushstring( L, argv[ i ] );
        lua_rawseti( L, -2, i );
    }
    lua_setglobal( L, "arg" );

    /* Open the executable to read the modules that are appended the file */
    FILE * const f = fopen( argv[ 0 ], "rb" );
    if( !f )
    {
        dogfood_errno( L );
    }

    const long payload_end_marker_length = sizeof( "\r\n-- dogfood DEADBEEF\r\n" ) - 1;
    if( fseek( f, -payload_end_marker_length, SEEK_END ) )
    {
        dogfood_errno( L );
    }

    const long end_of_payload = ftell( f );
    if( end_of_payload == -1L )
    {
        dogfood_errno( L );
    }

    unsigned long payload_offset = 0UL;
    if( fscanf( f, "\r\n-- dogfood %08lX\r\n", &payload_offset ) != 1 &&
        payload_offset == 0 )
    {
        dogfood_error( L, "No valid payload end marker found." );
    }

    if( fseek( f, ( long )payload_offset, SEEK_SET ) )
    {
        dogfood_errno( L );
    }

    /* The main module is the first module in the payload */
    module_name_buffer main_module_name = { 0 };
    unsigned long      main_module_size = 0UL;
    read_module_header( L, f, main_module_name, &main_module_size );

    const long main_module_pos = ftell( f );
    if( !( main_module_pos != -1L &&
           fseek( f, ( long )main_module_size, SEEK_CUR ) == 0 ) )
    {
        dogfood_errno( L );
    }

    /* Load modules (byte)code and add it to the package.loaded table */
    lua_getglobal( L, "package" );
    lua_getfield( L, -1, "loaded" );

    long pos = ftell( f );
    while( ( pos != -1L ) && ( pos < end_of_payload ) )
    {
        module_name_buffer module_name = { 0 };
        unsigned long      module_size = 0UL;

        read_module_header( L, f, module_name, &module_size );
        load_module( L, f, module_name, module_size );

        lua_call( L, 0, 1 );
        lua_setfield( L, -2, module_name );

        pos = ftell( f );
    }

    /* Load and execute the main module (byte)code */
    if( ferror( f ) ||
        fseek( f, main_module_pos, SEEK_SET ) )
    {
        dogfood_errno( L );
    }

    load_module( L, f, main_module_name, main_module_size );

    fclose( f );

    const int status = lua_pcall( L, 0, 1, 0 );

    if( status != LUA_OK )
    {
        /* Show Lua errors.
         * Do not prepend 'Dogfood error:' to the error message. The error has nothing
         * to do with dogfood since it is emited by the Lua VM or the Lua modules. */
        const char * const msg = lua_tostring( L, -1 );
        fprintf( stderr, "%s\n", msg ? msg : "An error has occurred." );
    }

    /* Get the exit status if the return value is a number */
    const int exit_status = ( int )( lua_isnumber( L, 0 ) ? lua_tonumber( L, 0 ) : 0.0 );

    lua_close( L );

    /* Return exit status when OK, else default to 1 when there was an error */
    return status == LUA_OK ? exit_status : 1;
}


src/food.lua
-- Copyright (c) 2019 PG1003
--
-- Permission is hereby granted, free of charge, to any person obtaining a copy
-- of this software and associated documentation files (the "Software"), to
-- deal in the Software without restriction, including without limitation the
-- rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
-- sell copies of the Software, and to permit persons to whom the Software is
-- furnished to do so, subject to the following conditions:

-- The above copyright notice and this permission notice shall be included in
-- all copies or substantial portions of the Software.

-- THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
-- IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
-- FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
-- AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
-- LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
-- OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
-- SOFTWARE.


local help = [[dogfood v1.1.0

A tool  for creating self running Lua programs.

Usage: dogfood [OPTIONS] OUT MODULE...

Creates a program at location OUT that runs its embedded Lua MODULEs.
The first MODULE from the list is the entry point of the program while
the optional extra modules are accessable by require. The modules are
provided without the '.lua' extention.
C modules cannot be embedded in the resulting program but the program can
load C modules when they are placed in one of the package.cpath search paths.

The following OPTIONS are available;
-c, --compile                   Compile the modules and embed them as bytecode.
-s, --strip-debug-information   Strips the debug information from the bytecode.
-h, --help                      Show this help.
-m                              Adds the given path to package.loaded to search
                                  for modules that are provided as parameter.
-v, --lua-version               Show the Lua language version of the
                                  interpreter used by the resulting program.]]


local function dogfood_error( msg )
    io.stderr:write( "Dogfood error: " .. msg .. "\n" )
    os.exit( 1 )
end

local function dogfood_warn( msg )
    io.stderr:write( "Dogfood warning: " .. msg .. "\n" )
end

--
-- Parse and validate the commandline parameters
--

local compile                 = false
local strip_debug_information = false
local out                     = false
local modules                 = {}
        
do
    if #arg == 0 then
        dogfood_error( "Use '-h' or '--help' as parameters for information about the usage of dogfood." )
    end

    local displayed_help        = false
    local displayed_lua_version = false

    local option = false
    for _, param in ipairs( arg ) do
        if option then
            if option == "-m" then
                local path   = param:gsub( "\"", "" )
                package.path = package.path .. ";" .. path
            end
            option = false
        elseif param == "-h" or param == "--help" then
            print( help )
            displayed_help = true
        elseif param == "-c" or param == "--compile" then
            compile = true
        elseif param == "-s" or param == "--strip-debug-information" then
            strip_debug_information = true
        elseif param == "-v" or param == "--lua-version" then
            print( _VERSION )
            displayed_lua_version = true
        elseif param == "-m" then
            option = param
        else
            if not out then
                out = param:gsub( "\"", "" )
            else
                modules[ #modules + 1 ] = { name = param }
            end
        end
    end

    -- '-h' and '-v' options display only information and do not require the OUT and MODULE parameters
    if ( displayed_help or displayed_lua_version ) and not out and #modules == 0 then
        os.exit( 0 )
    end

    if not out then
        dogfood_error( "No output file provided." )
    end

    if #modules == 0 then
        dogfood_error( "No modules provided." )
    end
end

--
-- Validate modules
--

do
    local found = {}
    local purge = {}
    for i, mod in ipairs( modules ) do
        if found[ mod.name ] then
            -- Duplicates
            purge[ #purge + 1 ] = i
        else
            local path = package.searchpath( mod.name, package.path )
            if path then
                mod.path = path
            else
                path = package.searchpath( mod.name, package.cpath )
                if path then
                    local ok, result = pcall( require, mod.name )
                    if ok then
                        dogfood_warn( "Module '" .. mod.name ..
                                      "' is a C module found at location '" ..
                                      path .. "'." )
                        -- C modules cannot be embedded
                        purge[ #purge + 1 ] = i
                    else
                        dogfood_error( result )
                    end
                else
                    dogfood_error( "Cannot find module '" .. mod.name .. "'." )
                end
            end
            found[ mod.name ] = true
        end
    end

    for i = #purge, 1, -1 do
        table.remove( modules, purge[ i ] )
    end
end

--
-- Open the output file to write the executable data to
--

local output = io.open( out, "w+b" )
if not output then
    dogfood_error( "Cannot create output file '" .. out .. "'." )
end

local payload_end = false

--
-- Copy the 'dog' part of this dogfood executable to the new executable
--

do
    local dogfood = io.open( arg[ 0 ], "rb" )
    if not dogfood then
        dogfood_error( "Cannot read own dogfood executable." )
    end

    if not dogfood:seek( "end", -23 ) then
        dogfood_error( "Cannot seek to end of the dogfood executable." )
    end

    payload_end = dogfood:read( 23 )
    if not payload_end then
        dogfood_error( "No end-of-payload marker found." )
    end

    local dog_bytes = payload_end:match( "%s-- dogfood (%x%x%x%x%x%x%x%x)%s" )
    if not dog_bytes then
        dogfood_error( "Invalid end-of-payload marker found." )
    end

    dog_bytes = assert( tonumber( dog_bytes, 16 ) )
    if dog_bytes >= ( dogfood:seek() - 23 ) then
        dogfood_error( "Invalid end-of-payload marker value." )
    end

    dogfood:seek( "set", 0 )
    repeat
        local count = dog_bytes > 4096 and 4096 or dog_bytes
        local chunk = assert( dogfood:read( count ) )
        output:write( chunk )
        dog_bytes = dog_bytes - count
    until dog_bytes <= 0
    dogfood:close()
end

--
-- Append payload
--

do
    for _, mod in ipairs( modules ) do
        -- Validate if the module has syntactic errors by loading the file
        local mod_chunk, error_msg = loadfile( mod.path, "t" );
        if mod_chunk == nil then
            dogfood_error( "Error while loading module.\n" .. error_msg )
        end

        local mod_data
        if compile then
            mod_data = assert( string.dump( mod_chunk, strip_debug_information ) )
        else
            -- Work around for 'read all' which has changed with 5.3
            local file = assert( io.open( mod.path, "rb" ) )
            file:seek( "end", 0 )
            local file_size = file:seek( "cur", 0 )
            file:seek( "set", 0 )
            mod_data = assert( file:read( file_size ) )
            file:close()
        end
        output:write( string.format( "\r\n-- %s %08X\r\n", mod.name, #mod_data ) )
        output:write( mod_data )
    end

    output:write( payload_end )
    output:close()
end

return 0
